### ISSUE TITLE
ocamlrun 5.2.0 fails on m68k

### ISSUE THREAD

**Issue by glondu** [ocaml/ocaml#13249]
OCaml 5.2.0 fails to build on Debian sid, m68k.

Quoting the [build log](https://buildd.debian.org/status/fetch.php?pkg=ocaml&arch=m68k&ver=5.2.0-1%7Eexp1&stamp=1718285451&raw=0):
```
  MKLIB runtime/libcamlrun.a
  MKEXE runtime/ocamlrun
cp runtime/ocamlrun boot/ocamlrun
/usr/bin/make -C stdlib OCAMLRUN='$(ROOTDIR)/boot/ocamlrun' USE_BOOT_OCAMLC=true all
make[3]: Entering directory '/<<PKGBUILDDIR>>/stdlib'
  OCAMLC camlinternalFormatBasics.cmi
qemu: uncaught target signal 11 (Segmentation fault) - core dumped
make[3]: *** [Makefile:121: camlinternalFormatBasics.cmi] Segmentation fault
make[3]: Leaving directory '/<<PKGBUILDDIR>>/stdlib'
make[2]: *** [Makefile:671: coldstart] Error 2
make[2]: Leaving directory '/<<PKGBUILDDIR>>'
```

It turns out the segfault happens on the following line:
```
 goto *(void *)(jumptbl_base + *pc++); /* Jump to the first instruction */
```
in `caml_interprete` (runtime/interp.c). In my setup, `jumptbl_base` is NULL (which looks suspicious) and `pc` is 0xc1aad008. Are there legit cases where this line is executed with `jumptbl_base` being NULL?

**Comment by nojb** [comment]
I believe that m68k is a 32-bit architecture, in which case it is normal for `jumptbl_base` to be 0, see:

https://github.com/ocaml/ocaml/blob/f1c28574accca8a8693b5097d32e0d1d4a19e906/runtime/interp.c#L56-L61

**Comment by glondu** [comment]
Yes, m68k is 32-bit and big-endian.

I could manage to go past the `goto` instruction, which turns out to lead to code related to the BRANCH bytecode instruction.

Before the `goto`, `caml_state` points to 0x8004c560 and has address 0xc0aa752c. After the `goto`, `caml_state` is NULL and has address 0xc0aae50e. I don't understand this change, the only thing specific to this variable is its thread_local-ness, but there is a single thread here... By making `caml_state` not thread_local (which is obviously wrong), I can see the build progress (but I suspect it will blow up later when multi-threaded code is involved).

**Comment by glondu** [comment]
> By making `caml_state` not thread_local (which is obviously wrong), I can see the build progress [...]

The build completes and, as expected, many tests (mostly parallel-related but not only) fail. I've attached [tests.log](https://github.com/user-attachments/files/15896021/tests.log) for reference.

Now one has to understand what is wrong with `caml_state` being thread_local...

**Comment by glondu** [comment]
I've asked the Debian m68k porters: https://lists.debian.org/debian-68k/2024/06/msg00000.html

**Comment by glondu** [comment]
The problem seems to be in the following lines:

https://github.com/ocaml/ocaml/blob/f1c28574accca8a8693b5097d32e0d1d4a19e906/runtime/interp.c#L213-L217

Register `a5` seems to be used for TLS. AFAIU, the consequence of the above lines is to use register `a5` as `pc` variable, but then every change to `pc` changes also `caml_state`. Neutralizing the lines seems to fix the issue... I'm waiting for a new build with this change to complete...

Is there a reason why register variables are pinned like this?

**Comment by ghost** [comment]
Good catch. This logic predates use of thread-local storage in OCaml and was used to speed-up things slightly. Given the scarce number of available registers on m68k, I think it would be best to remove that block entirely.

**Comment by ghost** [comment]
Turns out it's not TLS but simply PIC (i.e. all non-static binaries). Can you give the diff in #13252 a try?

**Comment by glondu** [comment]
> Can you give the diff in #13252 a try?

I tried a seemingly equivalent diff. The build succeeds, but 16 tests [fail](https://github.com/user-attachments/files/15902537/tests.log).

I had a look at the first one, and it turns out `Obj.tag (Obj.repr [| |])` always return 1002 (unaligned_tag). This might be related to [this](https://lists.debian.org/debian-68k/2024/06/msg00003.html).

The compiler and runtime work sufficiently to build the Debian package, which is enough for me. Fixing the testsuite is cherry on the cake.

**Comment by shindere** [comment]
Stéphane Glondu (2024/06/19 07:53 -0700):
> The compiler and runtime work sufficiently to build the Debian
> package, which is enough for me. Fixing the testsuite is cherry on the
> cake.

I got curious about this claim. To me a fialure in the testsuite means
that the compiler is not really working and I tend to take those
failures quite seriously. Aren't you of this opinion?

**Comment by glondu** [comment]
Le 21/06/2024 à 11:33, Sébastien Hinderer a écrit :
>  > The compiler and runtime work sufficiently to build the Debian
>  > package, which is enough for me. Fixing the testsuite is cherry on the
>  > cake.
> 
> I got curious about this claim. To me a fialure in the testsuite means
> that the compiler is not really working and I tend to take those
> failures quite seriously. Aren't you of this opinion?

In general, I agree. But I don't have the resources to do more myself.
